File: bindings/wasm/benchmarks/throughput.mjs
// Throughput benchmark for the wickra-terminal WASM binding.
//
// What this measures is the boundary, not the core. Every binding drives the
// same Rust terminal through one function -- a command JSON in, a frame JSON
// out -- so the number is the cost of crossing this boundary once per command.
// WASM is the interesting row: there is no FFI in the usual sense, but every
// string crossing is copied into and out of linear memory, which is a real cost
// the other bindings do not pay in the same shape.
//
// Build the nodejs target first:
//
//     wasm-pack build bindings/wasm --target nodejs --release --out-dir pkg-node
//
// then, from the repository root:
//
//     node bindings/wasm/benchmarks/throughput.mjs
//     node bindings/wasm/benchmarks/throughput.mjs --ticks 100000

import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);
const { Terminal } = require('../pkg-node/wickra_terminal_wasm.js');

// Shared by all nine binding benchmarks, so the numbers compare.
const CONFIG = JSON.stringify({
  sources: [{ Synth: { seed: 1 } }],
  layout: {
    panels: [
      { kind: 'Chart', rect: { x: 0, y: 0, w: 100, h: 40 } },
      { kind: 'Book', rect: { x: 0, y: 40, w: 50, h: 30 } },
      { kind: 'Tape', rect: { x: 50, y: 40, w: 50, h: 30 } },
    ],
  },
});
const SUBSCRIBE = JSON.stringify({ type: 'Subscribe', source: 0, symbol: 'BTC/USDT' });
const TICK = JSON.stringify({ type: 'Tick' });
const LIST = JSON.stringify({ type: 'ListIndicators' });

// The catalogue response is ~30 kB, so a hundred of them is a noisy sample.
const CATALOGUE_REPS = 1000;

/** Median elapsed nanoseconds over a few repetitions, after one warmup pass. */
function medianNs(run, reps = 3) {
  run();
  const samples = [];
  for (let i = 0; i < reps; i++) {
    const start = process.hrtime.bigint();
    run();
    samples.push(Number(process.hrtime.bigint() - start));
  }
  samples.sort((a, b) => a - b);
  return samples[Math.floor(samples.length / 2)];
}

const flag = process.argv.indexOf('--ticks');
const parsed = flag === -1 ? NaN : Number(process.argv[flag + 1]);
const ticks = Number.isFinite(parsed) && parsed >= 100 ? parsed : 20000;

const term = new Terminal(CONFIG);
term.command(SUBSCRIBE);
const frameBytes = Buffer.byteLength(term.command(TICK));
const catalogueBytes = Buffer.byteLength(term.command(LIST));

const tickNs = medianNs(() => {
  for (let i = 0; i < ticks; i++) term.command(TICK);
});
const listNs = medianNs(() => {
  for (let i = 0; i < CATALOGUE_REPS; i++) term.command(LIST);
});

const n = (value) => value.toLocaleString('en-US');
console.log(`wickra-terminal WASM throughput - ${n(ticks)} commands (median of 3)\n`);
console.log('Command'.padEnd(18) + 'per second'.padStart(14) + 'us/command'.padStart(14) + 'payload'.padStart(12));
console.log('-'.repeat(58));
console.log(
  'Tick'.padEnd(18) +
    n(Math.round(ticks / (tickNs / 1e9))).padStart(14) +
    (tickNs / ticks / 1e3).toFixed(2).padStart(14) +
    `${n(frameBytes)}B`.padStart(12),
);
console.log(
  'ListIndicators'.padEnd(18) +
    n(Math.round(CATALOGUE_REPS / (listNs / 1e9))).padStart(14) +
    (listNs / CATALOGUE_REPS / 1e3).toFixed(2).padStart(14) +
    `${n(catalogueBytes)}B`.padStart(12),
);
console.log(
  '\nOne command crosses the boundary once. Higher is better, and the numbers\n' +
    'are machine-dependent -- compare bindings on one machine, never across two.',
);

File: bindings/c/benchmarks/throughput.c
/* Throughput benchmark for the wickra-terminal C ABI.
 *
 * What this measures is the boundary, not the core. Every binding drives the
 * same Rust terminal through one function -- a command JSON in, a frame JSON
 * out -- so the number is the cost of crossing this boundary once per command.
 * C is the floor the others are measured against: there is no marshalling here
 * beyond passing two pointers, so whatever this costs is the command itself
 * plus one allocation for the returned string.
 *
 * The Rust core runs the identical loop with no boundary at all, as
 * examples/rust --bin throughput. That is the floor; this is the cheapest a
 * boundary gets.
 *
 * Built by bindings/c/benchmarks/CMakeLists.txt:
 *
 *     cargo build -p wickra-terminal-c --release
 *     cmake -S bindings/c/benchmarks -B bindings/c/benchmarks/build
 *     cmake --build bindings/c/benchmarks/build --config Release
 *     ./bindings/c/benchmarks/build/throughput [ticks]
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "wickra_terminal.h"

/* Shared by all nine binding benchmarks, so the numbers compare. */
static const char *const CONFIG =
    "{\"sources\":[{\"Synth\":{\"seed\":1}}],"
    "\"layout\":{\"panels\":["
    "{\"kind\":\"Chart\",\"rect\":{\"x\":0,\"y\":0,\"w\":100,\"h\":40}},"
    "{\"kind\":\"Book\",\"rect\":{\"x\":0,\"y\":40,\"w\":50,\"h\":30}},"
    "{\"kind\":\"Tape\",\"rect\":{\"x\":50,\"y\":40,\"w\":50,\"h\":30}}]}}";
static const char *const SUBSCRIBE =
    "{\"type\":\"Subscribe\",\"source\":0,\"symbol\":\"BTC/USDT\"}";
static const char *const TICK = "{\"type\":\"Tick\"}";
static const char *const LIST = "{\"type\":\"ListIndicators\"}";

/* The catalogue response is ~30 kB, so a hundred of them is a noisy sample. */
#define CATALOGUE_REPS 1000

static double now_ns(void) {
    struct timespec ts;
    timespec_get(&ts, TIME_UTC);
    return (double)ts.tv_sec * 1e9 + (double)ts.tv_nsec;
}

/* Drive `count` commands, freeing each frame. Returns 0 on success. */
static int drive(WickraTerminal *term, const char *command, long count) {
    for (long i = 0; i < count; i++) {
        char *out = NULL;
        const int status = wickra_terminal_command(term, command, &out);
        if (status != WICKRA_TERMINAL_OK) {
            fprintf(stderr, "command failed: %s\n", out ? out : command);
            wickra_terminal_free_string(out);
            return 1;
        }
        wickra_terminal_free_string(out);
    }
    return 0;
}

/* Median of three timed runs, after one warmup. */
static double median_ns(WickraTerminal *term, const char *command, long count) {
    double samples[3];
    if (drive(term, command, count) != 0) {
        return -1.0;
    }
    for (int rep = 0; rep < 3; rep++) {
        const double start = now_ns();
        if (drive(term, command, count) != 0) {
            return -1.0;
        }
        samples[rep] = now_ns() - start;
    }
    for (int i = 0; i < 2; i++) {
        for (int j = i + 1; j < 3; j++) {
            if (samples[j] < samples[i]) {
                const double swap = samples[i];
                samples[i] = samples[j];
                samples[j] = swap;
            }
        }
    }
    return samples[1];
}

/* The byte length of one response, for the payload column. */
static size_t payload(WickraTerminal *term, const char *command) {
    char *out = NULL;
    if (wickra_terminal_command(term, command, &out) != WICKRA_TERMINAL_OK) {
        wickra_terminal_free_string(out);
        return 0;
    }
    const size_t length = out ? strlen(out) : 0;
    wickra_terminal_free_string(out);
    return length;
}

int main(int argc, char **argv) {
    long ticks = argc > 1 ? strtol(argv[1], NULL, 10) : 20000;
    if (ticks < 100) {
        ticks = 20000;
    }

    WickraTerminal *term = wickra_terminal_new(CONFIG);
    if (!term) {
        fprintf(stderr, "could not build a terminal from the benchmark config\n");
        return 1;
    }
    if (drive(term, SUBSCRIBE, 1) != 0) {
        wickra_terminal_free(term);
        return 1;
    }

    const size_t frame_bytes = payload(term, TICK);
    const size_t catalogue_bytes = payload(term, LIST);
    const double tick_ns = median_ns(term, TICK, ticks);
    const double list_ns = median_ns(term, LIST, CATALOGUE_REPS);
    wickra_terminal_free(term);

    if (tick_ns < 0.0 || list_ns < 0.0) {
        return 1;
    }

    printf("wickra-terminal C throughput - %ld commands (median of 3)\n\n", ticks);
    printf("%-18s%14s%14s%12s\n", "Command", "per second", "us/command", "payload");
    printf("----------------------------------------------------------\n");
    printf("%-18s%14.0f%14.2f%11zuB\n", "Tick", (double)ticks / (tick_ns / 1e9),
           tick_ns / (double)ticks / 1e3, frame_bytes);
    printf("%-18s%14.0f%14.2f%11zuB\n", "ListIndicators", (double)CATALOGUE_REPS / (list_ns / 1e9),
           list_ns / (double)CATALOGUE_REPS / 1e3, catalogue_bytes);
    printf("\nOne command crosses the boundary once. Higher is better, and the numbers\n"
           "are machine-dependent -- compare bindings on one machine, never across two.\n");
    return 0;
}
